File: C#.Net/JDI/Web/Selenium/Elements/Complex/table/Columns.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading;
using Epam.JDI.Commons;
using Epam.JDI.Web.Selenium.Elements.Complex.table.interfaces;
using OpenQA.Selenium;
using static Epam.JDI.Core.ExceptionUtils;
using static Epam.JDI.Core.Settings.JDISettings;

namespace Epam.JDI.Web.Selenium.Elements.Complex.table
{
    public class Columns : TableLine
    {
        public Columns()
        {
            HasHeader = true;
            ElementIndex = ElementIndexType.Nums;
            HeadersLocator = By.XPath(".//th");
            DefaultTemplate = By.XPath(".//tr/td[{0}]");
        }

        protected override Func<TableLine, IList<IWebElement>> GetHeadersAction =>
            row =>
            {
                var headers = Table.WebElement.FindElements(HeadersLocator);
                return Table.Rows.SkipFirstColumn() ? headers.ListCopy(1) : headers;
            };

        protected override IList<IWebElement> GetFirstLine()
        {
            var line = Table.Rows.GetLineAction(1);
            return Table.Rows.SkipFirstColumn() ? line.ListCopy(1) : line;
        }
        
        public Dictionary<string, ICell> GetColumn(string colName)
        {
            return ActionWithException(() =>
            {
                var rowsCount = Table.Rows.Count;
                var headers = Table.Rows.Headers;
                var webColumn = Timer.GetResultByCondition(() => GetLineAction(colName), els => els.Count == rowsCount);
                var result = new Dictionary<string, ICell>();
                for (var i = 0; i < rowsCount; i++)
                    result.Add(Table.Rows.Headers[i], Table.Cell(webColumn[i], new Column(name: colName), new Row(name: headers[i])));
                return result;
            }, ex => $"Can't Get Column '{colName}'. Reason: {ex}");
        }

        public IList<string> GetColumnValue(string colName)
        {
            return
                ActionWithException(
                    () => Table.Columns.GetLineAction(colName).Select(el => el.Text).ToList(), 
                    ex => $"Can't Get Column '{colName}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetColumnAsText(string colName)
        {
            return GetColumn(colName).ToDictionary(pair => pair.Key, pair => pair.Value.Text);
        }

        public Dictionary<string, ICell> CellsToColumn(Collection<ICell> cells)
        {
            return cells.ToDictionary(cell => Headers[cell.RowNum - 1], cell => cell);
        }

        public Dictionary<string, ICell> GetColumn(int colNum)
        {
            if (Count < 0 || Table.Columns.Count < colNum || colNum <= 0)
                throw Exception($"Can't Get Column '{colNum}'. [num] > RowsCount({Count}).");
            return ActionWithException(() =>
            {
                var rowsCount = Table.Rows.Count;
                var webColumn = Timer.GetResultByCondition(() => GetLineAction(colNum),
                    els => els.Count == rowsCount);
                var result = new Dictionary<string, ICell>();
                for (var i = 0; i < rowsCount; i++)
                    result.Add(Table.Rows.Headers[i],
                        Table.Cell(webColumn[i], new Column(colNum), new Row(i + 1)));
                return result;
            }, ex => $"Can't Get Column '{colNum}'. Reason: {ex}");
        }

        public IList<string> GetColumnValue(int colNum)
        {
            if (Count < 0 || Count < colNum || colNum <= 0)
                throw Exception($"Can't Get Column '{colNum}'. [num] > RowsCount({Count}).");
            return ActionWithException(() => Table.Columns.GetLineAction(colNum).Select(el => el.Text).ToList(),
                ex => $"Can't Get Column '{colNum}'. Reason: {ex}");
        }

        public Dictionary<string, string> GetColumnAsText(int colNum)
        {
            return GetColumn(colNum).ToDictionary(pair => pair.Key, pair => pair.Value.Text);
        }

        public override Dictionary<string, Dictionary<string, ICell>> Get()
        {
            return Headers.ToDictionary(key =>key, GetColumn);
        }
        
    }
}

File: Java/JDI/jdi-uitest-win/src/main/java/com/epam/jdi/uitests/win/winium/elements/complex/table/Columns.java
package com.epam.jdi.uitests.win.winium.elements.complex.table;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.map.MapArray;
import com.epam.jdi.uitests.core.interfaces.common.IText;
import com.epam.jdi.uitests.win.winium.elements.complex.table.interfaces.ICell;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.Collection;
import java.util.List;

import static com.epam.commons.LinqUtils.select;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by 12345 on 26.10.2014.
 */
public class Columns extends TableLine {
    public Columns() {
        hasHeader = true;
        elementIndex = ElementIndexType.Nums;
        headersLocator = By.xpath(".//th");
        defaultTemplate = By.xpath(".//tr/td[%s]");
    }

    protected List<WebElement> getHeadersAction() {
        return table.getWebElement().findElements(headersLocator);
    }

    protected List<WebElement> getFirstLine() {
        return table.rows().getLineAction(1);
    }

    public final MapArray<String, ICell> getColumn(String colName) {
        try {
            int rowsCount = table.rows().count();
            List<String> headers = table.rows().headers();
            List<WebElement> webColumn = timer().getResultByCondition(
                    () -> getLineAction(colName), els -> els.size() == rowsCount);
            return new MapArray<>(rowsCount,
                    table.rows().headers::get,
                    value -> table.cell(webColumn.get(value), new Column(colName), new Row(headers.get(value))));
        } catch (Exception | Error ex) {
            throw throwColumnException(colName, ex.getMessage());
        }
    }

    public final List<String> getColumnValue(String colName) {
        try {
            return select(table.columns().getLineAction(colName), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwColumnException(colName, ex.getMessage());
        }
    }

    public final MapArray<String, String> getColumnAsText(String colName) {
        return getColumn(colName).toMapArray(IText::getText);
    }

    public MapArray<String, ICell> cellsToColumn(Collection<ICell> cells) {
        return new MapArray<>(cells,
                cell -> headers().get(cell.rowNum() - 1),
                cell -> cell);
    }

    public final MapArray<String, ICell> getColumn(int colNum) {
        if (count() < 0 || table.columns().count() < colNum || colNum <= 0)
            throw exception("Can't Get Column '%s'. [num] > RowsCount(%s).", colNum, count());
        try {
            int rowsCount = table.rows().count();
            List<WebElement> webColumn = timer().getResultByCondition(
                    () -> getLineAction(colNum), els -> els.size() == rowsCount);
            return new MapArray<>(rowsCount,
                    key -> table.rows().headers().get(key),
                    value -> table.cell(webColumn.get(value), new Column(colNum), new Row(value + 1)));
        } catch (Exception | Error ex) {
            throw throwColumnException(colNum + "", ex.getMessage());
        }
    }

    public final List<String> getColumnValue(int colNum) {
        if (count() < 0 || count() < colNum || colNum <= 0)
            throw exception("Can't Get Column '%s'. [num] > RowsCount(%s).", colNum, count());
        try {
            return select(table.columns().getLineAction(colNum), WebElement::getText);
        } catch (Exception | Error ex) {
            throw throwColumnException(colNum + "", ex.getMessage());
        }
    }

    public final MapArray<String, String> getColumnAsText(int colNum) {
        return getColumn(colNum).toMapArray(IText::getText);
    }

    public MapArray<String, MapArray<String, ICell>> get() {
        return new MapArray<>(headers(), key -> key, this::getColumn);
    }

    private RuntimeException throwColumnException(String columnName, String ex) {
        return exception("Can't Get Column '%s'. Reason: %s", columnName, ex);
    }
}